File: Hill/Hill.py
import numpy as np

def createMatrixFromKey(key, size):
    key_matrix = []
    for i in range(size):
        row = []
        for j in range(size):
            row.append(ord(key[i * size + j]) % 65)
        key_matrix.append(row)
    return np.array(key_matrix)

def encryptMessage(message, key_matrix, size):
    message_vector = []
    for i in range(size):
        message_vector.append(ord(message[i]) % 65)
    message_vector = np.array(message_vector)
    
    cipher_matrix = np.dot(key_matrix, message_vector) % 26
    cipher_text = ''.join(chr(int(cipher_matrix[i]) + 65) for i in range(size))
    
    return cipher_text

def hillCipherEncrypt(message, key):
    size = int(len(key) ** 0.5)
    key_matrix = createMatrixFromKey(key, size)
    
    message = message.upper().replace(" ", "")
    if len(message) % size != 0:
        message += 'X' * (size - len(message) % size)
    
    cipher_text = ''
    for i in range(0, len(message), size):
        cipher_text += encryptMessage(message[i:i+size], key_matrix, size)
    
    return cipher_text

# Example usage
try:
    message = input("Enter the message: ").upper().replace(" ", "")
    key = input("Enter the key: ").upper().replace(" ", "")
    if len(key) ** 0.5 != int(len(key) ** 0.5):
        raise ValueError("Key length must be a perfect square.")
except Exception as e:
    print(f"Error: {e}")
    exit(1)
encrypted_message = hillCipherEncrypt(message, key)
print(f"Encrypted Message: {encrypted_message}")

File: Hill/Hill.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

vector<vector<int>> GetKeyMatrix(const string& key, int size) {
    vector<vector<int>> keyMatrix(size, vector<int>(size));
    int k = 0;
    for (int i = 0; i < size; i++) {
        for (int j = 0; j < size; j++) {
            keyMatrix[i][j] = (key[k]) % 65;
            k++;
        }
    }
    return keyMatrix;
}

vector<int> GetMessageVector(const string& message, int size) {
    vector<int> messageVector(size);
    for (int i = 0; i < size; i++) {
        messageVector[i] = (message[i]) % 65;
    }
    return messageVector;
}

string EncryptMessage(const string& message, const vector<vector<int>>& keyMatrix, int size) {
    vector<int> messageVector = GetMessageVector(message, size);
    vector<int> cipherVector(size);

    for (int i = 0; i < size; i++) {
        cipherVector[i] = 0;
        for (int j = 0; j < size; j++) {
            cipherVector[i] += keyMatrix[i][j] * messageVector[j];
        }
        cipherVector[i] = cipherVector[i] % 26;
    }

    string cipherText;
    for (int i = 0; i < size; i++) {
        cipherText += cipherVector[i] + 65;
    }

    return cipherText;
}

int main() {
    string message, key;
    cout << "Enter the message: ";
    cin >> message;
    cout << "Enter the key: ";
    cin >> key;

    if (message.length() != 3 || key.length() != 9) {
        cerr << "Error: Message must be 3 characters long and key must be 9 characters long." << endl;
        return 1;
    }

    int size = 3; // Assuming a 3x3 key matrix for simplicity

    vector<vector<int>> keyMatrix = GetKeyMatrix(key, size);
    string cipherText = EncryptMessage(message, keyMatrix, size);

    cout << "Encrypted Message: " << cipherText << endl;

    return 0;
}